File: FastSubsetSum.py
import sys
import os


def fast_subset_sum(X, T):
    """
    FastSubsetSum algorithm from Erickson's Algorithms (2019).
    Uses a set-based dynamic programming approach.
    Returns True if there is a subset of X with sum exactly T.
    """
    S = {0}
    for x in X:
        S = S | {s + x for s in S}
        S = {s for s in S if s <= T}
    return T in S


def main():
    if len(sys.argv) < 2:
        print("Usage: python FastSubsetSum.py <input_file>")
        sys.exit(1)

    input_file = sys.argv[1]

    try:
        with open(input_file, 'r') as f:
            lines = f.read().split()
    except OSError as e:
        print(f"Error reading input file: {e}")
        sys.exit(1)

    try:
        n = int(lines[0])
        if len(lines) < n + 2:
            raise ValueError(f"Expected {n + 2} tokens, got {len(lines)}")
        X = [int(lines[i + 1]) for i in range(n)]
        T = int(lines[n + 1])
    except (ValueError, IndexError) as e:
        print(f"Error parsing input: {e}")
        sys.exit(1)

    result = fast_subset_sum(X, T)

    os.makedirs('outputs', exist_ok=True)

    input_filename = os.path.basename(input_file)
    output_filename = os.path.join('outputs', 'output ' + input_filename)

    with open(output_filename, 'w') as f:
        f.write('TRUE\n' if result else 'FALSE\n')

    print('TRUE' if result else 'FALSE')


if __name__ == '__main__':
    main()

File: FastSubsetSum.java
import java.io.*;
import java.util.*;

public class FastSubsetSum {

    /**
     * FastSubsetSum algorithm from Erickson's Algorithms (2019).
     * Uses a set-based dynamic programming approach.
     * Returns true if there is a subset of X with sum exactly T.
     */
    public static boolean fastSubsetSum(int[] X, int T) {
        Set<Integer> S = new HashSet<>();
        S.add(0);
        for (int x : X) {
            Set<Integer> newS = new HashSet<>(S);
            for (int s : S) {
                int val = s + x;
                if (val <= T) {
                    newS.add(val);
                }
            }
            S = newS;
        }
        return S.contains(T);
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.out.println("Usage: java FastSubsetSum <input_file>");
            System.exit(1);
        }

        String inputFile = args[0];

        Scanner scanner;
        try {
            scanner = new Scanner(new File(inputFile));
        } catch (FileNotFoundException e) {
            System.out.println("Error reading input file: " + e.getMessage());
            System.exit(1);
            return;
        }

        int n = scanner.nextInt();
        int[] X = new int[n];
        for (int i = 0; i < n; i++) {
            X[i] = scanner.nextInt();
        }
        int T = scanner.nextInt();
        scanner.close();

        boolean result = fastSubsetSum(X, T);

        new File("outputs").mkdirs();

        String inputFilename = new File(inputFile).getName();
        String outputFilename = "outputs/output " + inputFilename;

        PrintWriter writer = new PrintWriter(new FileWriter(outputFilename));
        writer.println(result ? "TRUE" : "FALSE");
        writer.close();

        System.out.println(result ? "TRUE" : "FALSE");
    }
}
